File: seoyeong/solved/LEET_1584.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        parent = list(range(n))

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(x, y):
            root_x, root_y = find(x), find(y)
            
            if root_x != root_y:
                parent[root_y] = root_x
                return True
            return False

        edges = []

        for i in range(n):
            for j in range(i+1, n):
                if i == j: continue

                [x1, y1], [x2, y2] = points[i], points[j]
                d = abs(x1 - x2) + abs(y1 - y2)
                edges.append([d, i, j])

        edges.sort()

        answer = 0
        cnt = 0

        for cost, i, j in edges:
            if union(i, j):
                answer += cost
                cnt += 1

                if cnt == n-1:
                    break
        
        return answer

File: kyungrin/solved/LEET_1584.java
import java.util.*;

public class LEET_1584 {

    class Node implements Comparable<Node> {
        int a, b;
        int val;

        public Node(int a, int b, int val) {
            this.a = a;
            this.b = b;
            this.val = val;
        }

        @Override
        public int compareTo(Node o) {
            return Integer.compare(this.val, o.val);
        }
    }

    int[] parent;
    PriorityQueue<Node> pq;

    public int minCostConnectPoints(int[][] points) {

        // 1. 간선 생성
        pq = new PriorityQueue<>();
        int cnt = points.length;

        if (cnt <= 1) return 0;

        for (int i = 0; i < cnt; i++) {
            for (int j = i + 1; j < cnt; j++) {
                int val =
                        Math.abs(points[i][0] - points[j][0])
                        + Math.abs(points[i][1] - points[j][1]);

                pq.add(new Node(i, j, val));
            }
        }

        // 2. 간선 선정
        int n = 0;
        int result = 0;

        parent = new int[cnt];

        for (int i = 0; i < cnt; i++) {
            parent[i] = i;
        }

        while (true) {
            Node cur = pq.poll();

            int a = cur.a;
            int b = cur.b;
            int val = cur.val;

            if (find(a) == find(b)) continue;

            union(a, b);
            result += val;
            n++;

            if (n == cnt - 1) break;
        }

        // 3. 출력
        return result;
    }

    public int find(int x) {
        if (x == parent[x]) return x;
        return parent[x] = find(parent[x]);
    }

    public void union(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        parent[rootB] = rootA;
    }
}
